### ISSUE TITLE
crypten.init() gets hung

### ISSUE THREAD

**Issue by JenySadadia** [facebookresearch/CrypTen#282]
Hi,

I am trying to use crypten and it gets hung on initialization.

>>>import crypten
>>>crypten.init()

I tried with torch.set_num_threads(1)  solution I found on github issues.
That gave me below error : 

>>> crypten.init()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/lib/python3.8/site-packages/crypten/__init__.py", line 21, in init
    if comm.get().get_rank() < comm.get().get_world_size():
  File "/lib/python3.8/site-packages/crypten/communicator/__init__.py", line 21, in get
    return cls.get()
  File "/lib/python3.8/site-packages/crypten/communicator/distributed_communicator.py", line 91, in get
    return cls.instance
AttributeError: type object 'DistributedCommunicator' has no attribute 'instance'

Development Environment:
I am using python 3.8.5
Please have a look at the modules I have installed (pip3 freeze output):

absl-py==0.11.0
aio-pika==6.8.0
aiofiles==0.6.0
aiohttp==3.6.3
aiormq==3.3.1
alembic==1.4.3
appdirs==1.4.4
APScheduler==3.7.0
asgiref==3.3.4
astunparse==1.6.3
async-generator==1.10
async-timeout==3.0.1
attrs==19.3.0
audioread==2.1.9
backcall==0.2.0
bidict==0.21.2
boto3==1.17.72
botocore==1.20.72
cachetools==4.2.2
certifi==2020.12.5
cffi==1.14.5
chardet==3.0.4
click==8.0.0
cloudpickle==1.6.0
colorclass==2.2.0
coloredlogs==14.3
colorhash==1.0.3
coreapi==2.3.3
coreschema==0.0.4
croniter==1.0.13
crypten==0.1
cryptography==3.4.7
cycler==0.10.0
dataclasses==0.6
decorator==5.0.7
deepspeech==0.6.1
Django==3.1.7
django-rest-passwordreset==1.1.0
django-rest-swagger==2.2.0
django-storages==1.11.1
djangorestframework==3.12.2
djangorestframework-simplejwt==4.6.0
dm-tree==0.1.6
dnspython==1.16.0
docopt==0.6.2
fbmessenger==6.0.0
filelock==3.0.12
future==0.18.2
gast==0.3.3
gitdb==4.0.7
GitPython==3.1.17
google-api-core==1.26.3
google-api-python-client==2.4.0
google-auth==1.30.0
google-auth-httplib2==0.1.0
google-auth-oauthlib==0.4.4
google-cloud-core==1.6.0
google-cloud-storage==1.37.0
google-crc32c==1.1.2
google-pasta==0.2.0
google-resumable-media==1.2.0
googleapis-common-protos==1.53.0
grpcio==1.37.1
gTTS==2.2.2
h11==0.9.0
h5py==2.10.0
httpcore==0.11.1
httplib2==0.19.1
httptools==0.2.0
httpx==0.15.4
huggingface-hub==0.0.12
humanfriendly==9.1
HyperPyYAML==1.0.0
idna==2.10
ipython-genutils==0.2.0
isodate==0.6.0
itypes==1.2.0
jedi==0.18.0
Jinja2==3.0.0
jmespath==0.10.0
joblib==0.15.1
jsonpickle==1.5.2
jsonschema==3.2.0
kafka-python==1.4.7
Keras-Preprocessing==1.1.2
kiwisolver==1.3.1
librosa==0.8.0
llvmlite==0.36.0
Mako==1.1.4
Markdown==3.3.4
MarkupSafe==2.0.0
matplotlib==3.4.1
matplotlib-inline==0.1.2
mattermostwrapper==2.2
multidict==4.7.6
mysqlclient==2.0.3
networkx==2.5.1
numba==0.53.1
numpy==1.21.0
oauth2client==4.1.3
oauthlib==3.1.0
onnx==1.9.0
openapi-codec==1.3.2
opt-einsum==3.3.0
packaging==20.9
pamqp==2.3.0
parso==0.8.2
pexpect==4.8.0
pickleshare==0.7.5
pika==1.2.0
Pillow==8.2.0
pooch==1.3.0
prompt-toolkit==2.0.10
protobuf==3.17.0
psycopg2-binary==2.8.6
ptyprocess==0.7.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pydot==1.4.2
Pygments==2.9.0
PyJWT==2.1.0
pykwalify==1.8.0
pymongo==3.10.1
pyparsing==2.4.7
pyrsistent==0.17.3
pyTelegramBotAPI==3.7.7
python-crfsuite==0.9.7
python-dateutil==2.8.1
python-editor==1.0.4
python-engineio==4.1.0
python-http-client==3.3.2
python-socketio==5.2.1
python-speech-features==0.6
pytz==2020.5
PyWavelets==1.1.1
PyYAML==5.4.1
questionary==1.5.2
rasa-sdk==2.6.0
rasa-x==0.39.3
redis==3.5.3
regex==2020.9.27
requests==2.25.1
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
resampy==0.2.2
rfc3986==1.5.0
rocketchat-API==1.9.1
rsa==4.7.2
ruamel.yaml==0.16.13
ruamel.yaml.clib==0.2.2
s3transfer==0.4.2
sanic==20.12.3
Sanic-Cors==0.10.0.post3
sanic-jwt==1.6.0
Sanic-Plugins-Framework==0.9.5
scikit-learn==0.24.2
scipy==1.7.0
sendgrid==6.6.0
sentencepiece==0.1.96
sentry-sdk==0.19.5
simplejson==3.17.2
six==1.16.0
sklearn-crfsuite==0.3.6
slackclient==2.9.3
smmap==4.0.0
sniffio==1.2.0
SoundFile==0.10.3.post1
speechbrain==0.5.9
SpeechRecognition==3.8.1
SQLAlchemy==1.3.24
sqlparse==0.4.1
starkbank-ecdsa==1.1.0
stt==0.10.0a6
tabulate==0.8.9
tensorboard==2.5.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.0
tensorflow-addons==0.12.0
tensorflow-estimator==2.3.0
tensorflow-hub==0.10.0
tensorflow-probability==0.11.1
tensorflow-text==2.3.0
termcolor==1.1.0
terminaltables==3.1.0
threadpoolctl==2.1.0
torch==1.9.0
torchaudio==0.9.0
torchvision==0.5.0
tqdm==4.61.1
traitlets==5.0.5
twilio==6.50.1
typeguard==2.12.0
typing-extensions==3.10.0.0
tzlocal==2.1
ujson==1.35
uritemplate==3.0.1
urllib3==1.26.4
uvloop==0.14.0
wcwidth==0.2.5
webexteamssdk==1.6
websockets==8.1
Werkzeug==2.0.0
wrapt==1.12.1
yarl==1.5.1

Looking forward to get a solution from the community.
Thanks.

**Comment by knottb** [comment]
Hard to tell exactly what is happening here. If `crypten.init()` is hanging, it is usually because `torch.distributed` is waiting for a quorum to initialize (i.e. all parties involved in the computation must call `crypten.init()`). I am assuming you are trying to do run this with a single party here?

Perhaps check if you have the right number of parties specified - run the following:

```
import os

print(os.environ.get("WORLD_SIZE"], 1))
print(os.environ.get("RANK", 0))
```

If you are running with 1 party, these should print `1` and `0` respectively.

It's unclear to me why you would get an instance doesn't exist error though. In either case, it means that torch.distributed did not initialize as expected. Let me know if you have any additional information.

**Comment by JenySadadia** [comment]
Hello,

Thank you for your response.
I tried to check the parties and below is the result I got:

>>> print(os.environ.get("WORLD_SIZE", 1))
1
>>> print(os.environ.get("RANK",0))
0

Please tell me if you need any other information to debug this issue.
Perhaps a version-specific issue is there because when I was trying to run crypten with torch 1.4.0 , It worked fine.
The issue occurred after installing torch 1.9.0.

I would be glad if you can help.

**Comment by knottb** [comment]
I'm not exactly sure what would be causing this issue. I think we have seen a few issues pop up around PyTorch 1.9.0, so perhaps try using an earlier version of PyTorch for the time being. We will try to fix these issues to work with the newest PyTorch version moving forward.

**Comment by knottb** [comment]
Is this still causing issues. We have released crypten version 0.4.0 which should resolve any issues related to PyTorch versioning.
